File: WiFiDirect/js/js/scenario2_connector.js
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

(function () {
    "use strict";
    var page = WinJS.UI.Pages.define("/html/scenario2_connector.html", {
        ready: function (element, options) {
            document.getElementById("btnGetDevices").addEventListener("click", btnGetDevices_Click, false);
            document.getElementById("btnIe").addEventListener("click", btnIe_Click, false);
            document.getElementById("btnFromId").addEventListener("click", btnFromId_Click, false);
            document.getElementById("btnSendMessage").addEventListener("click", btnSendMessage_Click, false);
            document.getElementById("btnCloseDevice").addEventListener("click", btnCloseDevice_Click, false);

            var discoveredDevicesList = document.getElementById("lvDiscoveredDevices").winControl;
            discoveredDevicesListBinding = new WinJS.Binding.List([]);

            discoveredDevicesList.itemDataSource = discoveredDevicesListBinding.dataSource;
            discoveredDevicesList.forceLayout();

            var connectedDevicesList = document.getElementById("lvConnectedDevices").winControl;
            connectedDevicesListBinding = new WinJS.Binding.List([]);

            connectedDevicesList.itemDataSource = connectedDevicesListBinding.dataSource;
            connectedDevicesList.forceLayout();
        }
    });

    var strServerPort = "50001";

    var customOui = [170, 187, 204];
    var customOuiType = 221;
    var wfaOui = [80, 111, 154];
    var msftOui = [0, 80, 242];

    var discoveredDevicesListBinding;
    var connectedDevicesListBinding;

    function btnGetDevices_Click() {
        WinJS.log && WinJS.log("Finding Devices......", "sample", "status");

        discoveredDevicesListBinding.splice(0, discoveredDevicesListBinding.length);

        var deviceSelector;
        switch (cmbDeviceSelector.value) {
            case "AssociationEndpoint":
                deviceSelector = Windows.Devices.WiFiDirect.WiFiDirectDevice.getDeviceSelector(Windows.Devices.WiFiDirect.WiFiDirectDeviceSelectorType.associationEndpoint);
                break;

            default:
                deviceSelector = Windows.Devices.WiFiDirect.WiFiDirectDevice.getDeviceSelector(Windows.Devices.WiFiDirect.WiFiDirectDeviceSelectorType.deviceInterface);
                break;
        }

        var targetWfdProperties = ["System.Devices.WiFiDirect.InformationElements"];
        Windows.Devices.Enumeration.DeviceInformation.findAllAsync(deviceSelector, targetWfdProperties).done(
            function (deviceInfoCollection) {
                if (deviceInfoCollection.size === 0) {
                    WinJS.log && WinJS.log("No devices found", "sample", "error");
                    return;
                }

                deviceInfoCollection.forEach(function (item, idx) {
                    discoveredDevicesListBinding.push(new WfdHelper.DiscoveredDevice(item));
                });

                WinJS.log && WinJS.log(deviceInfoCollection.size.toString() + " devices found", "sample", "status");
            });
    }

    function btnIe_Click() {
        var discoveredDevicesList = document.getElementById("lvDiscoveredDevices").winControl;

        if (discoveredDevicesList.selection.count() === 0) {
            WinJS.log && WinJS.log("No device selected, please select one", "sample", "error");
            return;
        }

        discoveredDevicesList.selection.getItems().done(function (items) {
            var deviceInfo = items[0].data.deviceInfo;
            var iECollection = Windows.Devices.WiFiDirect.WiFiDirectInformationElement.createFromDeviceInformation(deviceInfo);

            if (iECollection.size === 0) {
                WinJS.log && WinJS.log("No Information Elements found", "sample", "error");
                return;
            }

            var strIeOutput = "";
            var strIe = "N/A";

            iECollection.forEach(function (ie) {
                var oui = new Uint8Array(ie.oui.length);
                var dataReader = Windows.Storage.Streams.DataReader.fromBuffer(ie.oui);
                dataReader.readBytes(oui);
                dataReader.close();

                var bOuiStr = oui.toString(16);
                if (bOuiStr === msftOui.toString()) {
                    strIeOutput = strIeOutput.concat("Microsoft IE: ");
                }
                else if (bOuiStr === wfaOui.toString()) {
                    strIeOutput = strIeOutput.concat("WFA IE: ");
                }
                else if (bOuiStr === customOui.toString()) {
                    strIeOutput = strIeOutput.concat("Custom IE: ");

                    var dataReader = Windows.Storage.Streams.DataReader.fromBuffer(ie.value);
                    dataReader.unicodeEncoding = Windows.Storage.Streams.UnicodeEncoding.utf8;
                    dataReader.byteOrder = Windows.Storage.Streams.ByteOrder.littleEndian;

                    strIe = dataReader.readString(dataReader.readUInt32());
                }
                else {
                    strIeOutput = strIeOutput.concat("Unknown IE: ");
                }

                strIeOutput = strIeOutput + "OUI Type: " + ie.ouiType.toString() + " OUI: " + bOuiStr + " IE Data: " + strIe + "\n";
            });

            WinJS.log && WinJS.log(strIeOutput, "sample", "status");
        });
    }

    function btnFromId_Click() {
        var discoveredDevicesList = document.getElementById("lvDiscoveredDevices").winControl;

        if (discoveredDevicesList.selection.count() === 0) {
            WinJS.log && WinJS.log("No device selected, please select one", "sample", "error");
            return;
        }


        discoveredDevicesList.selection.getItems().done(function (items) {
            var deviceInfo = items[0].data.deviceInfo;

            WinJS.log && WinJS.log("Connecting to " + deviceInfo.name + "...", "sample", "status");

            var connectionParams = new Windows.Devices.WiFiDirect.WiFiDirectConnectionParameters();
            connectionParams.groupOwnerIntent = parseInt(txtGOIntent.value.toString());

            Windows.Devices.WiFiDirect.WiFiDirectDevice.fromIdAsync(deviceInfo.id, connectionParams).then(function (wfdDevice) {
                if (wfdDevice === null) {
                    WinJS.log && WinJS.log("Connection to " + deviceInfo.name + " failed", "sample", "error");
                    return;
                }

                // Register for the ConnectionStatusChanged event handler
                wfdDevice.onconnectionstatuschanged = OnConnectionStatusChanged;

                // Get the EndpointPair information.
                var endpointPairCollection = wfdDevice.getConnectionEndpointPairs();

                WinJS.log && WinJS.log("Devices connected on L2 layer, connecting to IP Address: " + endpointPairCollection[0].remoteHostName +
                " Port: " + strServerPort, "sample", "status");

                // Wait for server to start listening on a socket
                var wakeupTime = new Date().getTime() + 2000;
                while (new Date().getTime() < wakeupTime);

                // Connect to Advertiser on L4 layer
                var clientSocket = new Windows.Networking.Sockets.StreamSocket();
                var endPointPair = endpointPairCollection[0];
                endPointPair.remoteServiceName = strServerPort;
                clientSocket.connectAsync(endPointPair).done(function () {
                    var sessionId = Math.random().toString();

                    var socketRW = new WfdHelper.SocketReaderWriter(clientSocket, sessionId);
                    connectedDevicesListBinding.push(new WfdHelper.ConnectedDevice(sessionId, wfdDevice, socketRW));

                    socketRW.ReadMessage();
                    socketRW.WriteMessage(sessionId);

                    WinJS.log && WinJS.log("Connected with remote side on L4 layer", "sample", "status");
                },
                function (err) {
                    WinJS.log && WinJS.log("Socket connection failed: " + err.toString(), "sample", "error");
                    return;
                });
            },
            function (err) {
                WinJS.log && WinJS.log("Connect operation failed: " + err.toString(), "sample", "error");
                return;
            });
        });
    }

    function OnConnectionStatusChanged(e) {
        WinJS.log && WinJS.log("Connection disconnected", "sample", "status");
    }

    function btnSendMessage_Click() {
        var connectedDevicesList = document.getElementById("lvConnectedDevices").winControl;

        if (connectedDevicesList.selection.count() == 0) {
            WinJS.log && WinJS.log("Please select a Session to send data", "sample", "error");
            return;
        }

        var messsage = document.getElementById("txtSendMessage");
        if (messsage.value === "") {
            WinJS.log && WinJS.log("Please type a message to send", "sample", "error");
            return;
        }

        connectedDevicesList.selection.getItems().done(function (items) {
                items[0].data.socketRW.WriteMessage(messsage.value);
            },
        function (err) {
            WinJS.log && WinJS.log("Send failed: " + +err.toString(), "sample", "error");
        });
    }

    function btnCloseDevice_Click() {
        var connectedDevicesList = document.getElementById("lvConnectedDevices").winControl;

        if (connectedDevicesList.selection.count() == 0) {
            WinJS.log && WinJS.log("Please select a device to close", "sample", "error");
            return;
        }

        connectedDevicesList.selection.getItems().done(function (items) {
            if (items[0].data.socketRW !== null) items[0].data.socketRW.Dispose();
            if (items[0].data.wfdDevice !== null) items[0].data.wfdDevice.close();
            connectedDevicesListBinding.splice(items[0].index, 1);
        },
        function (err) {
            WinJS.log && WinJS.log("Close failed: " + +err.toString(), "sample", "error");
        });
    }
})();

File: Samples/WiFiDirect/cppwinrt/Scenario2_Connector.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario2_Connector.h"
#include "SampleConfiguration.h"
#include "SocketReaderWriter.h"
#include "Scenario2_Connector.g.cpp"
#include <chrono>
#include <sstream>

using namespace std::chrono_literals;
using namespace winrt;
using namespace Windows::Devices::Enumeration;
using namespace Windows::Devices::WiFiDirect;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Networking;
using namespace Windows::Networking::Sockets;
using namespace Windows::Security::Cryptography;
using namespace Windows::Storage::Streams;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario2_Connector::Scenario2_Connector()
    {
        InitializeComponent();
    }

    void Scenario2_Connector::OnNavigatedFrom(NavigationEventArgs const&)
    {
        if (deviceWatcher)
        {
            StopWatcher();
        }
    }

    void Scenario2_Connector::StopWatcher()
    {
        deviceWatcher.Added(watcherAddedToken);
        deviceWatcher.Removed(watcherRemovedToken);
        deviceWatcher.Updated(watcherUpdatedToken);
        deviceWatcher.EnumerationCompleted(watcherEnumerationCompletedToken);
        deviceWatcher.Stopped(watcherStoppedToken);

        deviceWatcher.Stop();
        deviceWatcher = nullptr;
    }

    void Scenario2_Connector::btnWatcher_Click(IInspectable const&, RoutedEventArgs const&)
    {
        if (!deviceWatcher)
        {
            publisher.Start();

            if (publisher.Status() != WiFiDirectAdvertisementPublisherStatus::Started)
            {
                rootPage.NotifyUser(L"Failed to start advertisement.", NotifyType::ErrorMessage);
                return;
            }

            discoveredDevices.Clear();
            rootPage.NotifyUser(L"Finding Devices...", NotifyType::StatusMessage);

            hstring deviceSelector = WiFiDirectDevice::GetDeviceSelector(
                GetSelectedItemTag<WiFiDirectDeviceSelectorType>(cmbDeviceSelector()));

            deviceWatcher = DeviceInformation::CreateWatcher(deviceSelector, { L"System.Devices.WiFiDirect.InformationElements" });

            watcherAddedToken = deviceWatcher.Added({ get_weak(), &Scenario2_Connector::OnDeviceAdded });
            watcherRemovedToken = deviceWatcher.Removed({ get_weak(), &Scenario2_Connector::OnDeviceRemoved });
            watcherUpdatedToken = deviceWatcher.Updated({ get_weak(), &Scenario2_Connector::OnDeviceUpdated });
            watcherEnumerationCompletedToken = deviceWatcher.EnumerationCompleted({ get_weak(), &Scenario2_Connector::OnEnumerationCompleted });
            watcherStoppedToken = deviceWatcher.Stopped({ get_weak(), &Scenario2_Connector::OnStopped });

            deviceWatcher.Start();

            btnWatcher().Content(box_value(L"Stop Watcher"));
        }
        else
        {
            publisher.Stop();

            btnWatcher().Content(box_value(L"Start Watcher"));

            StopWatcher();

            rootPage.NotifyUser(L"Device watcher stopped.", NotifyType::StatusMessage);
        }
    }

    #pragma region DeviceWatcherEvents
    fire_and_forget Scenario2_Connector::OnDeviceAdded(DeviceWatcher const&, DeviceInformation deviceInfo)
    {
        auto lifetime = get_strong();
        co_await resume_foreground(Dispatcher());
        discoveredDevices.Append(make<DiscoveredDevice>(deviceInfo));
    }

    fire_and_forget Scenario2_Connector::OnDeviceRemoved(DeviceWatcher const&, DeviceInformationUpdate deviceInfoUpdate)
    {
        auto lifetime = get_strong();
        co_await resume_foreground(Dispatcher());
        unsigned int size = discoveredDevices.Size();
        for (unsigned int index = 0; index < size; index++)
        {
            SDKTemplate::DiscoveredDevice discoveredDevice = discoveredDevices.GetAt(index);
            DiscoveredDevice* discoveredDeviceImpl = get_self<DiscoveredDevice>(discoveredDevice);
            if (discoveredDeviceImpl->DeviceInfo().Id() == deviceInfoUpdate.Id())
            {
                discoveredDevices.RemoveAt(index);
                break;
            }
        }
    }

    fire_and_forget Scenario2_Connector::OnDeviceUpdated(DeviceWatcher const&, DeviceInformationUpdate deviceInfoUpdate)
    {
        auto lifetime = get_strong();
        co_await resume_foreground(Dispatcher());
        for (SDKTemplate::DiscoveredDevice const& discoveredDevice : discoveredDevices)
        {
            DiscoveredDevice* discoveredDeviceImpl = get_self<DiscoveredDevice>(discoveredDevice);
            if (discoveredDeviceImpl->DeviceInfo().Id() == deviceInfoUpdate.Id())
            {
                discoveredDeviceImpl->UpdateDeviceInfo(deviceInfoUpdate);
                break;
            }
        }
    }

    void Scenario2_Connector::OnEnumerationCompleted(DeviceWatcher const&, IInspectable const&)
    {
        rootPage.NotifyUser(L"DeviceWatcher enumeration completed", NotifyType::StatusMessage);
    }

    void Scenario2_Connector::OnStopped(DeviceWatcher const&, IInspectable const&)
    {
        rootPage.NotifyUser(L"DeviceWatcher stopped", NotifyType::StatusMessage);
    }
    #pragma endregion

    void Scenario2_Connector::btnIe_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto discoveredDevice = lvDiscoveredDevices().SelectedItem().as<SDKTemplate::DiscoveredDevice>();

        IVector<WiFiDirectInformationElement> informationElements = nullptr;
        try
        {
            DiscoveredDevice* discoveredDeviceImpl = get_self<DiscoveredDevice>(discoveredDevice);
            informationElements = WiFiDirectInformationElement::CreateFromDeviceInformation(discoveredDeviceImpl->DeviceInfo());
        }
        catch (hresult_error const& ex)
        {
            rootPage.NotifyUser(L"No Information element found: " + ex.message(), NotifyType::ErrorMessage);
            return;
        }

        std::wostringstream message;

        for (WiFiDirectInformationElement const& informationElement : informationElements)
        {
            hstring ouiName = CryptographicBuffer::EncodeToHexString(informationElement.Oui());
            hstring value;

            com_array<uint8_t> ouiArray;
            CryptographicBuffer::CopyToByteArray(informationElement.Oui(), ouiArray);
            array_view<uint8_t const> ouiConstArray{ ouiArray.begin(), ouiArray.end() };

            if (ouiConstArray == array_view<uint8_t const>(MsftOui))
            {
                // The format of Microsoft information elements is documented here:
                // https://msdn.microsoft.com/en-us/library/dn392651.aspx
                // with errata here:
                // https://msdn.microsoft.com/en-us/library/mt242386.aspx
                ouiName = ouiName + L" (Microsoft)";
            }
            else if (ouiConstArray == array_view<uint8_t const>(WfaOui))
            {
                ouiName = ouiName + L" (WFA)";
            }
            else if (ouiConstArray == array_view<uint8_t const>(CustomOui))
            {
                ouiName = ouiName + L" (Custom)";

                if (informationElement.OuiType() == CustomOuiType)
                {
                    DataReader dataReader = DataReader::FromBuffer(informationElement.Value());
                    dataReader.UnicodeEncoding(UnicodeEncoding::Utf8);
                    dataReader.ByteOrder(ByteOrder::LittleEndian);

                    // Read the string.
                    try
                    {
                        hstring data = dataReader.ReadString(dataReader.ReadUInt32());
                        value = L"Data: " + data;
                    }
                    catch (hresult_error const&)
                    {
                        value = L"(Unable to parse)";
                    }
                }
            }

            message << L"OUI " << std::wstring_view(ouiName)
                << L", Type " << static_cast<uint32_t>(informationElement.OuiType())
                << L" " << std::wstring_view(value) << std::endl;
        }

        message << L"Information elements found: " << informationElements.Size();

        rootPage.NotifyUser(message.str(), NotifyType::StatusMessage);
    }

    fire_and_forget Scenario2_Connector::btnFromId_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        auto discoveredDevice = lvDiscoveredDevices().SelectedItem().as<SDKTemplate::DiscoveredDevice>();
        DeviceInformation deviceInfo = get_self<DiscoveredDevice>(discoveredDevice)->DeviceInfo();
        rootPage.NotifyUser(L"Connecting to " + deviceInfo.Name() + L"...", NotifyType::StatusMessage);

        if (!deviceInfo.Pairing().IsPaired())
        {
            if (!co_await connectionSettingsPanel().RequestPairDeviceAsync(deviceInfo.Pairing()))
            {
                co_return;
            }
        }

        WiFiDirectDevice wfdDevice = nullptr;
        try
        {
            // IMPORTANT: FromIdAsync needs to be called from the UI thread
            wfdDevice = co_await WiFiDirectDevice::FromIdAsync(deviceInfo.Id());
        }
        catch (hresult_canceled const&)
        {
            rootPage.NotifyUser(L"FromIdAsync was canceled by user", NotifyType::ErrorMessage);
            co_return;
        }

        // Register for the ConnectionStatusChanged event handler
        wfdDevice.ConnectionStatusChanged({ get_weak(), &Scenario2_Connector::OnConnectionStatusChanged });

        IVectorView<EndpointPair> endpointPairs = wfdDevice.GetConnectionEndpointPairs();
        HostName remoteHostName = endpointPairs.GetAt(0).RemoteHostName();

        rootPage.NotifyUser(L"Devices connected on L2 layer, connecting to IP Address: " + remoteHostName.DisplayName() +
            L" Port: " + ServerPortString, NotifyType::StatusMessage);

        // Wait for server to start listening on a socket
        co_await resume_after(2s);

        // resume_after does not preserve thread context. Get back to the UI thread.
        co_await resume_foreground(Dispatcher());

        // Connect to Advertiser on L4 layer
        StreamSocket clientSocket;
        try
        {
            co_await clientSocket.ConnectAsync(remoteHostName, ServerPortString);
            rootPage.NotifyUser(L"Connected with remote side on L4 layer", NotifyType::StatusMessage);
        }
        catch (hresult_error const& ex)
        {
            rootPage.NotifyUser(L"Connect operation threw an exception: " + ex.message(), NotifyType::ErrorMessage);
            co_return;
        }

        auto socketRW = make_self<SocketReaderWriter>(clientSocket, rootPage);

        hstring sessionId = to_hstring(CryptographicBuffer::GenerateRandomNumber());
        connectedDevices.Append(make<ConnectedDevice>(sessionId, wfdDevice, socketRW));

        // The first message sent over the socket is the name of the connection.
        co_await socketRW->WriteMessageAsync(sessionId);

        while (co_await socketRW->ReadMessageAsync() != nullptr)
        {
            // Keep reading messages
        }
    }

    void Scenario2_Connector::OnConnectionStatusChanged(WiFiDirectDevice const& sender, IInspectable const&)
    {
        rootPage.NotifyUser(L"Connection status changed: " + to_hstring(sender.ConnectionStatus()), NotifyType::StatusMessage);
    }

    fire_and_forget Scenario2_Connector::btnSendMessage_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();
        auto connectedDevice = lvConnectedDevices().SelectedItem().as<SDKTemplate::ConnectedDevice>();
        co_await get_self<ConnectedDevice>(connectedDevice)->SocketRW()->WriteMessageAsync(txtSendMessage().Text());
    }

    void Scenario2_Connector::btnClose_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto connectedDevice = lvConnectedDevices().SelectedItem().as<SDKTemplate::ConnectedDevice>();
        RemoveByValue(connectedDevices, connectedDevice);

        // Close socket and WiFiDirect object
        connectedDevice.Close();
    }

    fire_and_forget Scenario2_Connector::btnUnpair_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        auto discoveredDevice = lvDiscoveredDevices().SelectedItem().as<SDKTemplate::DiscoveredDevice>();
        DeviceInformation deviceInfo = get_self<DiscoveredDevice>(discoveredDevice)->DeviceInfo();

        DeviceUnpairingResult result = co_await deviceInfo.Pairing().UnpairAsync();
        rootPage.NotifyUser(L"Unpair result: " + to_hstring(result.Status()), NotifyType::StatusMessage);
    }
}
